README.md
MSISDN Gateway
==============

[![Build Status](https://travis-ci.org/mozilla-services/msisdn-gateway.svg?branch=master)](https://travis-ci.org/mozilla-services/msisdn-gateway)

This is a proof of concept of an MSISDN Gateway server that takes a phone number and
register it using an SMS validation mechanism.

[API docs](API.md)

[Mailing list](https://mail.mozilla.org/listinfo/loop-services-dev)

[Dummy client](https://github.com/ferjm/msisdn-verifier-client)


Registration process flow
-------------------------

  1. The client sends a request to ``/register``.

  -- The server chooses the verification process based on MSISDN, MCC and MNC
     codes and returns a session token and a verify endpoint.

  2. The client sends a request to ``/sms/verify`` (the verify endpoint).

  -- The server sends a SMS text message containing a pin code, and returns the
     number that was used to send it (the phone number of the server, useful
     for silent SMS catch)

  3. The client sends a request to verify the pin code (to ``/sms/verify_code``)
     with the session token and the pin code and gets back a BrowserID
     certificate.
  4. If needed, the client can also ask for a new pin code using the
     ``/sms/resend_code`` URL and its sessionToken.
  5. Finally the client can unregister itself using the ``/unregister`` URL
     and its sessionToken.


What is the needed stack?
-------------------------

<img src="http://www.gliffy.com/go/publish/image/5799498/L.png" />


How to install?
---------------

You will need to have redis-server installed:

### Linux

    apt-get install redis-server

### OS X

Assuming you have brew installed, use it to install redis:

    brew install redis

If you need to restart it (after configuration update):

    brew services restart redis

### All Platforms

Then clone the loop server and install its dependencies:

    git clone https://github.com/mozilla-services/msisdn-gateway.git
    cd msisdn-gateway && make install


How to run it?
--------------

You can create your configuration file in `config/{NODE_ENV}.json`

You need to generate the BrowserId keys by running `./bin/generate-keypair` and
add them to your configuration file.

`development` is the default environment.

    make runserver

is equivalent to:

    NODE_ENV=development make runserver


How to run the tests?
---------------------

To run DynamoDB tests you will need to install and run [fake_dynamo](https://github.com/ananthakumaran/fake_dynamo) or [ddbmock](https://pypi.python.org/pypi/ddbmock)

FakeDynamo needs ruby, ruby-dev(el) and rubygems.

    make test


How to update translation?
--------------------------

    make update-l10n

And for automated scripts:

    make update-l10n NOVERIFY=yes


Estimate Redis Memory Usage
---------------------------

    usage = nbUsers * 1216 + 600000 (bytes)

 - For 3.5M users 4 GB
 - For 10M users 12 GB

The biggest AWS Elasticache Redis virtual machine is 68GB large so if
we want to handle more that 60M users we will probably want to do some
sharding to have one redis for MSISDN validation and another one for
hawkSession management.

 - 600000 bytes is the size of an empty redis-server.
 - 1216 bytes is the fixed size of a given user including:
    - It's hawkSessionToken
    - It's code validation related data
    - It's MSISDN related data

The /unregister endpoint drops everything about a user.


Where to report bugs?
---------------------

You should report bugs/issues or feature requests via [Github Issues](https://github.com/mozilla-services/msisdn-gateway/issues)


License
-------

The MSISDN Gateway code is released under the terms of the
[Mozilla Public License v2.0](http://www.mozilla.org/MPL/2.0/). See the
`LICENSE` file at the root of the repository.


CHANGELOG
CHANGELOG
=========

0.7.0 (unreleased)
------------------

- No changes yet.


0.6.0 (2015-05-15)
------------------

- [bug]         1164404 — The heartbeat didn't detect the readonly redis cluster.


0.5.1 (2014-11-17)
------------------

- [bug]         lastVerifiedAt was missing


0.5.0 (2014-11-17)
------------------

 - [refactoring] SMS /sms/momt/ url update changed.
                 This version needs a SMS Provider configuration.
 - [feature]     1096102 — Upgrade jwcrypto to browserid-crypto
 - [feature]     1088105 — Add the lastVerifiedAt certificate parameter.


0.4.5 (2014-08-20)
------------------

- [bug] duration is in seconds and was used as being in milliseconds.


0.4.4 (2014-08-19)
------------------

- [bug] MCC set to null is considered as missing.


0.4.3 (2014-08-13)
------------------

- [bug] Fix from parameter for nexmo sendSms.


0.4.2 (2014-08-12)
------------------

- [feature] Replace status code 200 with empty JSON to code 204 - No Content
- [feature] Handle the MCC on sms/mt/verify in order to choose
            the right MtSender number.
- [bug] fix CORS to allow credentials


0.4.1 (2014-07-16)
------------------

- [feature] 1038676 — Disable MOMT flow if no specific or default number is set
- [feature] Add the --fd argument to let circus spawn the socket
- [feature] Add the /api-specs url for videur
- [bug]     1037604 — Really fix the decrypt error
- [config]  Update configuration example
- [feature] Install fake_dynamo only if needed


0.4.0 (2014-07-15)
------------------

- [feature] Add BeepSend SMS Gateway
- [feature] Add SMS Gateway fallback mechanism
- [feature] Moved to AWS-SDK to handle STS support
- [bug]     Make the loadtest Python 2.6 compatible


0.3.1 (2014-07-08)
------------------

- [bug]     Fix MSISDN certificate generation
- [feature] Improve loadtests debugging
- [bug]     Change loadtests configuration from duration to hits


0.3.0 (2014-07-03)
------------------

- [feature] 1031517 — Add loadtests
- [feature] Remove res.on("header") deprecation
- [feature] Graceful shutdown of the server ok SIGTERM
- [feature] 1032270 — Manage l10n translation
- [bug]     Handle Hawk auth behind and HTTPS ELB
- [feature] Add a DynamoDB persistent storage backend
- [feature] Handle https and http agents configuration
- [feature] Store encrypted MSISDN number into the database


0.2.0 (2014-06-06)
------------------

- [api-change] Add the SMS MOMT protocol
- [api-change] Rewrite the API flows
                 - POST /discover
                 - POST /register
                 - POST /unregister
                 - POST /sms/mt/verify
                 - SMS /sms/momt/verify
                 - POST /sms/verify_code
                 - POST /certificat/sign
- [feature]    Use the FxA error format
- [feature]    Use Hawk Session
- [feature]    Implement BrowserID IdP well-known info.
- [feature]    Allow a limited number of retries for a verification code
- [feature]    Return the right moVerifier wrt the MCC/MNC code
- [feature]    Manage Hawk session expiration


0.1.0 (2014-05-06)
------------------

- [feature] Get a first PoC working
- [feature] Add the SMS MT protocol
              - /register
              - /unregister
              - /sms/verify
              - /sms/verify_code
              - /sms/resend_code


loadtests/loadtest.py
import hashlib
import hmac
import json
import math
import os
import random
import time
from urlparse import urlparse

import mohawk
from requests.auth import AuthBase

from loads.case import TestCase

OMXEN_URL = os.getenv("OMXEN_URL", "http://omxen.dev.mozaws.net")

print "USING %s OMXEN endpoint" % OMXEN_URL

PERCENTAGE_OF_MT_FLOW = 50  # Remining are MOMT flows
PERCENTAGE_OF_WRONG_CODES = 34  # Remining are valid ones.
PERCENTAGE_OF_SHORT_CODES = 50  # Remining are right ones.
MAX_OMXEN_TIMEOUT = 2  # Seconds to poll from omxen.

PUBLIC_KEY = '{"algorithm":"DS","y":"2bb69e89652d97315f156852c5d1a36c7ebf0fb991a3f812e55c804ca8fa8a09e6f1880de99fcaf4f5417a78a88591679219acda438f9ef4d5afb9d0ae20216414822c6187ec4f180eae6f1f2d2a0f55095e24141c8a5f12344ce4d6319042c168c962de1029a9ac1a528da2d27e4f9550b7c203572a034f077480d9f6e3ef3968902f7547a7374c5e7475dcc81002af3308d09ace22124719a83451cbd872403d6f72cf0b13b5ae0c092a3bf0daf8a26df7d618bc731306ccfbe431c9a30206327bedd0adac35dd5623d017c07aa7fbb991c9364335278b80f1a979d5141fb6cf6cc92849bc6ca5e3c24d25e75f828c0efd57214e37be92d32f2c2208527f20","p":"d6c4e5045697756c7a312d02c2289c25d40f9954261f7b5876214b6df109c738b76226b199bb7e33f8fc7ac1dcc316e1e7c78973951bfc6ff2e00cc987cd76fcfb0b8c0096b0b460fffac960ca4136c28f4bfb580de47cf7e7934c3985e3b3d943b77f06ef2af3ac3494fc3c6fc49810a63853862a02bb1c824a01b7fc688e4028527a58ad58c9d512922660db5d505bc263af293bc93bcd6d885a157579d7f52952236dd9d06a4fc3bc2247d21f1a70f5848eb0176513537c983f5a36737f01f82b44546e8e7f0fabc457e3de1d9c5dba96965b10a2a0580b0ad0f88179e10066107fb74314a07e6745863bc797b7002ebec0b000a98eb697414709ac17b401","q":"b1e370f6472c8754ccd75e99666ec8ef1fd748b748bbbc08503d82ce8055ab3b","g":"9a8269ab2e3b733a5242179d8f8ddb17ff93297d9eab00376db211a22b19c854dfa80166df2132cbc51fb224b0904abb22da2c7b7850f782124cb575b116f41ea7c4fc75b1d77525204cd7c23a15999004c23cdeb72359ee74e886a1dde7855ae05fe847447d0a68059002c3819a75dc7dcbb30e39efac36e07e2c404b7ca98b263b25fa314ba93c0625718bd489cea6d04ba4b0b7f156eeb4c56c44b50e4fb5bce9d7ae0d55b379225feb0214a04bed72f33e0664d290e7c840df3e2abb5e48189fa4e90646f1867db289c6560476799f7be8420a6dc01d078de437f280fff2d7ddf1248d56e1a54b933a41629d6c252983c58795105802d30d7bcd819cf6ef"}'  # NOQA


class TestMSISDN(TestCase):
    omxen_url = OMXEN_URL

    def test_all(self):
        # Discover
        self.discover()

        # Create a token
        self.register()

        # Use the MO flow 50% of the time and the MT flow the remaining
        if random.randint(0, 100) < PERCENTAGE_OF_MT_FLOW:
            # 1. Ask MSISDN validation
            resp = self.start_mt_flow()
            self.assertEqual(resp.status_code, 204,
                             "Start MT Flow failed: %s (%s)" % (
                                 resp.content, resp.status_code))
            self.incr_counter("status-204")

        else:
            # 2. Send SMS /sms/momt/verify hawkId
            resp = self.start_momt_flow()
            self.assertEqual(resp.status_code, 200,
                             "Start MOMT Flow failed: %s (%s)" % (
                                 resp.content, resp.status_code))
            self.incr_counter("status-200")

        # Poll omxen for the message
        message = self.read_message()

        # Get the message code
        if random.randint(0, 100) > PERCENTAGE_OF_WRONG_CODES:
            # 1. Try to validate a valid code
            self.incr_counter("try-right-code")
            resp = self.verify_code(message)
            if resp.status_code == 200:
                # If it was a valid code generate a certificate
                self.incr_counter("status-200")
                self.incr_counter("ask-for-certificate")
                self.sign_certificate()
            else:
                # If we didn't validate the code from the omxen
                # message it is probably because two test where using
                # the same MSISDN at the same time
                self.assertEquals(resp.status_code, 400,
                                  "Omxen collision failed: %s (%s)" % (
                                      resp.content, resp.status_code))
                self.incr_counter("status-400")
                self.incr_counter("omxen-collision-PROBABLY-WRONG")
        else:
            # 2. Try to validate a wrong code
            self.incr_counter("try-wrong-code")
            resp = self.verify_code()
            self.assertEquals(resp.status_code, 400,
                              "Try wrong code failed: %s (%s)" % (
                                  resp.content, resp.status_code))
            self.incr_counter("status-400")

        # Unregister
        self.unregister()

    def __get_random_msisdn(self):
        code = "%d" % random.randint(0, 99999999)
        code = code.zfill(8)
        return "+336%s" % code

    def discover(self):
        resp = self.session.post(
            self.server_url + '/discover',
            data=json.dumps({"mcc": "204"}),
            headers={'Content-type': 'application/json'}
        )
        self.assertEquals(resp.status_code, 200,
                          "Discover endpoint failed: %s (%s)" % (
                              resp.content, resp.status_code))
        self.incr_counter("status-200")

    def register(self):
        resp = self.session.post(self.server_url + '/register')
        self.assertEquals(resp.status_code, 200,
                          "Register endpoint failed: %s (%s)" % (
                              resp.content, resp.status_code))
        self.incr_counter("status-200")
        self.incr_counter("register")
        try:
            sessionToken = resp.json()['msisdnSessionToken']
        except ValueError:
            self.fail("No JSON has been returned: %s (%s)" % (
                resp.content, resp.status_code))
        except KeyError:
            self.fail("msisdnSessionToken not found in response: %s" %
                      resp.content)

        self.hawk_auth = HawkAuth(self.server_url, sessionToken)

    def start_mt_flow(self):
        self.msisdn = self.__get_random_msisdn()
        self.shortVerificationCode = \
            random.randint(0, 100) < PERCENTAGE_OF_SHORT_CODES

        self.incr_counter("mt-flow")

        return self.session.post(
            self.server_url + '/sms/mt/verify',
            data=json.dumps({
                "msisdn": self.msisdn,
                "mcc": "208",
                "shortVerification": self.shortVerificationCode
            }),
            headers={'Content-type': 'application/json'},
            auth=self.hawk_auth)

    def start_momt_flow(self):
        self.msisdn = self.__get_random_msisdn()
        self.shortVerificationCode = False

        self.incr_counter("momt-flow")

        # You don't need to be authenticated to revoke a token.
        return self.session.get(
            self.server_url + '/sms/momt/',
            params={"provider": "nexmo",
                    "msisdn": self.msisdn.lstrip("+"),
                    "text": "/sms/momt/verify %s" % self.hawk_auth.hawk_id})

    def read_message(self):
        resp = self.session.get(self.omxen_url + '/receive',
                                params={"to": self.msisdn.lstrip("+")})
        try:
            messages = resp.json()
            self.assertTrue(isinstance(messages, list),
                            "Wrong JSON from OMXEN: %s" % messages)
        except ValueError:
            print resp.content
            raise

        start_time = time.time()

        #  Poll on the omxen message list for this number
        while (len(messages) < 1 and
               time.time() - start_time < MAX_OMXEN_TIMEOUT):
            time.sleep(1)
            resp = self.session.get(self.omxen_url + '/receive',
                                    params={"to": self.msisdn.lstrip("+")})
            try:
                messages = resp.json()
                self.assertTrue(isinstance(messages, list),
                                "Wrong JSON from OMXEN: %s" % messages)
            except ValueError:
                print resp.content
                raise

        if len(messages) >= 1:
            return messages[0]["text"]

    def verify_code(self, message=None):
        # Extract the code from the message
        if message is None:
            # No message, build a fake code
            code = "%d" % random.randint(0, 999999)
            code = code.zfill(6)
        else:
            if len(message) == 64:
                # Long Verification code
                # The message is the code
                code = message
            else:
                # Short verification code, the code is the last word.
                code = message.split()[-1]
        return self.session.post(self.server_url + "/sms/verify_code",
                                 data=json.dumps({"code": code}),
                                 headers={'Content-type': 'application/json'},
                                 auth=self.hawk_auth)

    def sign_certificate(self):
        data = {
            'duration': 10,
            'publicKey': PUBLIC_KEY
        }
        resp = self.session.post(self.server_url + '/certificate/sign',
                                 data=json.dumps(data),
                                 headers={'Content-type': 'application/json'},
                                 auth=self.hawk_auth)
        self.assertEqual(resp.status_code, 200,
                         "Sign certificate failed: %s (%s)" % (
                             resp.content, resp.status_code))
        self.incr_counter("status-200")

    def unregister(self):
        resp = self.session.post(self.server_url + '/unregister',
                                 headers={'Content-type': 'application/json'},
                                 auth=self.hawk_auth)
        self.incr_counter("unregister")
        self.assertEqual(resp.status_code, 204,
                         "Unregister failed: %s (%s)" % (resp.content,
                                                         resp.status_code))
        self.incr_counter("status-204")


def HKDF_extract(salt, IKM, hashmod=hashlib.sha256):
    """HKDF-Extract; see RFC-5869 for the details."""
    if salt is None:
        salt = b"\x00" * hashmod().digest_size
    return hmac.new(salt, IKM, hashmod).digest()


def HKDF_expand(PRK, info, L, hashmod=hashlib.sha256):
    """HKDF-Expand; see RFC-5869 for the details."""
    digest_size = hashmod().digest_size
    N = int(math.ceil(L * 1.0 / digest_size))
    assert N <= 255
    T = b""
    output = []
    for i in xrange(1, N + 1):
        data = T + info + chr(i)
        T = hmac.new(PRK, data, hashmod).digest()
        output.append(T)
    return b"".join(output)[:L]


def HKDF(secret, salt, info, size, hashmod=hashlib.sha256):
    """HKDF-extract-and-expand as a single function."""
    PRK = HKDF_extract(salt, secret, hashmod)
    return HKDF_expand(PRK, info, size, hashmod)


class HawkAuth(AuthBase):
    def __init__(self, server_url, tokendata):
        hawk_session = tokendata.decode('hex')
        self.server_url = server_url
        keyInfo = 'identity.mozilla.com/picl/v1/sessionToken'
        keyMaterial = HKDF(hawk_session, "", keyInfo, 32*3)
        self.hawk_id = keyMaterial[:32].encode("hex")
        self.auth_key = keyMaterial[32:64].encode("hex")
        self.credentials = {
            'id': self.hawk_id,
            'key': self.auth_key,
            'algorithm': 'sha256'
        }

    def __call__(self, r):
        r.headers['Host'] = urlparse(self.server_url).netloc
        sender = mohawk.Sender(
            self.credentials,
            r.url,
            r.method,
            content=r.body or '',
            content_type=r.headers.get('Content-Type', '')
        )

        r.headers['Authorization'] = sender.request_header
        return r
